### ISSUE TITLE
Multi-symbol dataset selectors: array-based training/validation/test

### ISSUE THREAD

**Issue by jeromeetienne** [jeromeetienne/transformer_bitcoin_ai#16]
## Context

Experiment 07 (fine-tuning Chronos/TimesFM on BTCUSDT 4h) is data-limited: a
single symbol over a few years caps how much signal the fine-tune can learn.
The goal is to scale training data by pulling in multiple symbols.

To express that cleanly, the dataset YAML changes from "one source + three
date ranges" to **three top-level arrays** — \`training\`, \`validation\`, \`test\`
— where each element is a CSV selector with its own \`market\`, \`symbol\`,
\`interval\`, \`period\`, \`start\`, \`end\`.

The change is shared across all eight experiments because they all use the
same loader (\`src/btc_ai/data/dataset.py\`) **and** the same downstream
consumption pattern (concat splits → one Darts \`TimeSeries\` → walk-forward
backtest → scalar metric). Both layers need to move together.

User decisions (confirmed):

- Breaking change across all 8 experiments — no Shape-A back-compat path.
- \`test\` is also an array; metrics aggregated across symbols (one summary
  number per run, plus per-symbol breakdown in \`metrics.json\`).
- Per-series experiments (naive, moving average, ARIMA, gradient boosting)
  fit one model per test symbol against the matching-symbol training
  selector and aggregate test metrics.
- \`interval\` is a per-selector field but the loader enforces uniformity
  across every selector in the YAML.

## New dataset YAML shape

Single shape — the existing Shape-A and Shape-B branches are both replaced.
Inline comments mandatory per project convention (every field annotated; no
abbreviations).

\`\`\`yaml
training:
  - market: spot           # Binance market: 'spot' | 'futures/um' | 'futures/cm'
    symbol: BTCUSDT        # Binance trading pair
    interval: 4h           # kline interval (must match every other selector in this file)
    period: monthly        # Binance Vision archive granularity: 'monthly' or 'daily'
    start: 2023-01-01      # UTC, inclusive
    end:   2024-08-01      # UTC, exclusive
validation:
  - market: spot
    symbol: BTCUSDT
    interval: 4h
    period: monthly
    start: 2024-08-01
    end:   2024-10-01
test:
  - market: spot
    symbol: BTCUSDT
    interval: 4h
    period: monthly
    start: 2024-10-01
    end:   2024-12-01
\`\`\`

Adding ETHUSDT to training is a one-element append to the \`training\` list;
nothing else changes in the YAML.

### \`src/btc_ai/data/dataset.py\`

Replace \`DatasetSplits\` with:

\`\`\`python
@dataclass(frozen=True)
class SelectorFrame:
        df: pd.DataFrame      # tz-aware DatetimeIndex, OHLCV columns
        symbol: str           # 'BTCUSDT'
        market: str           # 'spot' | 'futures/um' | 'futures/cm'

@dataclass(frozen=True)
class DatasetSplits:
        train: list[SelectorFrame]
        validation: list[SelectorFrame]
        test: list[SelectorFrame]
        interval: str         # uniform across every selector
\`\`\`

\`DatasetSplits.symbol_train\` / \`symbol_test\` are **removed** (breaking).
Callers that need symbols read them off each \`SelectorFrame\`.

Replace \`_load_shape_a\` and \`_load_shape_b\` with:

- \`_load_selector(req: KlineRequest, cache_dir: Path) -> SelectorFrame\` —
  one CSV via \`BinanceVisionLoader\`, sliced by \`[start, end)\`.
- \`_validate_uniform_interval(splits) -> str\` — raises if any selector
  deviates; returns the single interval.
- \`_check_non_empty\` and \`_check_ordered_disjoint\` generalized to operate
  per-symbol across the three split arrays.

Add a helper used by per-series experiments:

\`\`\`python
def group_selectors_by_symbol(splits: DatasetSplits) -> dict[tuple[str, str], DatasetSplits]:
        # key = (market, symbol); value = a DatasetSplits scoped to that symbol.
        # Raises ValueError if a test (market, symbol) has no training match.
\`\`\`

### \`src/btc_ai/data/__init__.py\`

Re-export \`SelectorFrame\`. Keep \`load_dataset_from_experiment_cfg\` signature.

### \`src/btc_ai/eval/aggregate.py\` (new)

\`\`\`python
@dataclass(frozen=True)
class PerSymbolResult:
        symbol: str
        market: str
        n_test_rows: int
        metrics: dict[str, float]
        predictions: pd.DataFrame   # close, pred, ref, strategy_return, ...

def aggregate_metrics(results: list[PerSymbolResult]) -> dict[str, float]:
        # Macro mean across symbols for every scalar metric.
        # NaNs propagate (do not silently drop a symbol).

def write_aggregated_metrics_json(
        results: list[PerSymbolResult],
        out_path: Path,
        interval: str,
        dataset: str,
        experiment: str,
) -> None:
        # { aggregate: {...}, per_symbol: [...], interval, dataset, experiment }
\`\`\`

Aggregation rule is **unweighted macro mean** across symbols. Length-weighted
would let one long test series dominate and defeat the multi-symbol point.
Per-symbol breakdown stays in \`metrics.json\` for leaderboard inspection.

### Sequence models (\`04_lstm\`, \`05_transformer\`, \`06_pretrained\`, \`07_pretrained_finetune\`)

Native multi-series via Darts:

1. Convert each \`SelectorFrame\` to a \`TimeSeries\` (one per selector).
2. \`model.fit(series=train_ts_list, val_series=val_ts_list)\` — Darts accepts
   \`Sequence[TimeSeries]\` out of the box.
3. For evaluation, loop over \`test_ts_list\`; call \`historical_forecasts\` per
   series; build one \`PerSymbolResult\` per test series; hand to
   \`aggregate_metrics\`.

For experiment 07 specifically, \`train_and_evaluate\` in
\`experiments/07_pretrained_finetune/run.py\` today does
\`pd.concat([train, val, test])\` → one \`target_full_ts\`. Restructure to:

- Per training/validation symbol: build \`(scaler, scaled_train_ts, scaled_val_ts)\`.
- Fit once with multi-series lists.
- Per test symbol: re-scale with that symbol's scaler, run walk-forward,
  compute per-symbol metrics, append a \`PerSymbolResult\`.
- Best-val checkpoint logic is unchanged (one model, multi-series fit).

### Per-series models (\`01_baseline_naive\`, \`01b_moving_average\`, \`02_arima\`, \`03_gradient_boosting\`)

Loop pattern via \`group_selectors_by_symbol\`:

\`\`\`python
results: list[PerSymbolResult] = []
for (market, symbol), per_symbol_splits in group_selectors_by_symbol(splits).items():
        # per_symbol_splits.train/validation/test are single-element lists
        # for this (market, symbol).
        result = fit_and_evaluate_one_symbol(per_symbol_splits, cfg)
        results.append(result)
metrics = aggregate_metrics(results)
write_aggregated_metrics_json(results, ...)
\`\`\`

Matching rule: for each test selector, find training selectors with the same
\`(market, symbol)\`. Concatenate those (sorted by \`start\`) into the single
training frame for that symbol; same for validation. A training selector
whose \`(market, symbol)\` does not appear in \`test\` is **ignored** (logged at
WARNING). A test \`(market, symbol)\` with no training match raises
\`ValueError\` from \`group_selectors_by_symbol\`.

## Config migration

Migrate all three existing YAMLs to the new shape (one-element arrays
preserve current behavior):

- \`configs/datasets/btc_4h_2024.dataset.yaml\`
- \`configs/datasets/btc_4h_2020_2024.dataset.yaml\`
- \`configs/datasets/btc_1h_2024.dataset.yaml\`

Rewrite \`configs/datasets/README.md\` — the
"two valid shapes" section becomes one shape with multi-selector arrays.

## Touch-point list (dependency order)

1. \`src/btc_ai/data/dataset.py\` — schema, loader, helpers.
2. \`src/btc_ai/data/__init__.py\` — re-export \`SelectorFrame\`.
3. **new** \`src/btc_ai/eval/aggregate.py\` — \`PerSymbolResult\`, \`aggregate_metrics\`, writer.
4. \`configs/datasets/*.dataset.yaml\` (3 files) — migrate.
5. \`configs/datasets/README.md\` — rewrite.
6. \`scripts/fetch_data.py\` — iterate selectors, per-symbol row counts.
7. \`experiments/01_baseline_naive/run.py\` — per-symbol loop.
8. \`experiments/01b_moving_average/run.py\` — same.
9. \`experiments/02_arima/run.py\` (+ \`sweep.py\`) — same.
10. \`experiments/03_gradient_boosting/run.py\` (+ \`sweep.py\`) — same.
11. \`experiments/04_lstm/run.py\` (+ \`sweep.py\`, \`optuna_sweep.py\`) — Darts list pattern.
12. \`experiments/05_transformer/run.py\` (+ sweeps) — same.
13. \`experiments/06_pretrained/run.py\` (+ \`sweep.py\`) — same.
14. \`experiments/07_pretrained_finetune/run.py\` (+ \`sweep.py\`) — biggest change; restructure \`train_and_evaluate\`.
15. Grep for \`symbol_train\` / \`symbol_test\` / \`splits.train.index\` / \`pd.concat([splits.train, splits.validation, splits.test])\` across all experiments and remove every reference.

## Verification

End-to-end smoke test — no full retraining required:

- \`uv run python scripts/fetch_data.py --config experiments/01_baseline_naive/configs/btc_4h_2024.config.yaml\` — confirms the migrated single-selector YAML returns the same row counts as before.
- \`uv run python experiments/01_baseline_naive/run.py --config experiments/01_baseline_naive/configs/btc_4h_2024.config.yaml\` — aggregate \`mae\` must equal the pre-refactor value to within float epsilon (single-symbol regression check).
- \`uv run python experiments/02_arima/run.py --config experiments/02_arima/configs/btc_4h_2024.config.yaml\` — same single-symbol equivalence check.
- Hand-author a throwaway \`configs/datasets/btc_4h_2024_multi.dataset.yaml\` that adds ETHUSDT to \`training\` (validation and test stay BTCUSDT-only). Re-run \`01_baseline_naive\` against it; aggregate must equal the BTCUSDT-only run, because the extra ETHUSDT training selector is ignored by the per-series matching rule. This proves the matching path runs without altering single-symbol semantics.
- \`uv run python experiments/07_pretrained_finetune/run.py --config experiments/07_pretrained_finetune/configs/btc_4h_2020_2024_encoder_only.chronos-2-small.config.yaml\` with \`n_epochs\` reduced for the smoke run — verifies multi-series fit path works on the smallest fine-tune backend.
- \`uv run make lint\` from the repo root — catches dead references to \`symbol_train\` / \`symbol_test\`.

## Critical files

- \`src/btc_ai/data/dataset.py\`
- \`src/btc_ai/eval/aggregate.py\` (new)
- \`configs/datasets/btc_4h_2024.dataset.yaml\`
- \`experiments/02_arima/run.py\` — canonical per-series template
- \`experiments/07_pretrained_finetune/run.py\` — canonical multi-series template
